### PR TITLE
Sort and Sample over Block

### LINKED ISSUE #34: Implement Sort and Sample over `Block`
- [ ] Implement `Key` with Ordered.

### PR DESCRIPTION
**Pull request by hataehyeok**
closes #34

### CODE REVIEW

File: core/src/main/scala/Block.scala [hunk 1]
```diff
@@ -30,6 +30,14 @@ object Block {
       valueLength: Int = 90
   ): Block =
     Block.fromSource(Source.fromURI(path.toURI), keyLength, valueLength)
+
+  def sort(block: Block): Block = {
+    val sortedRecords = block.records.sortBy(_.key)
+    new Block(sortedRecords)
+  }
```
Comment (betarixm) [inline core/src/main/scala/Block.scala:37]: Nit: add new line between methods.
```suggestion
  }

```

File: core/src/main/scala/Block.scala [hunk 2]
```diff
@@ -30,6 +30,14 @@ object Block {
...
+  def sample(block: Block): List[Key] = {
+    block.records.map(_.key).toList
```
Comment (betarixm) [inline core/src/main/scala/Block.scala:39]: In our design, `sample` method is in charge of sampling keys. E.g. not return *whole keys* of block.

File: core/src/main/scala/Key.scala [hunk 3]
```diff
@@ -8,7 +8,12 @@ class Key(val underlying: Array[Byte]) extends AnyVal with Ordered[Key] {
...
-    // TODO: Implement compare
-    underlying.length
+    val result = underlying
+      .zip(that.underlying)
+      .map { case (a, b) =>
+        a - b
+      }
+      .find(_ != 0)
+    result.getOrElse(0)
   }
```
Comment (betarixm) [inline core/src/main/scala/Key.scala:18]: Unneccesary code block and the variable.

```scala
  override def compare(that: Key): Int = underlying
    .zip(that.underlying)
    .map { case (a, b) =>
      a - b
    }
    .find(_ != 0)
    .getOrElse(0)
```

File: core/src/main/scala/Block.scala [hunk 4]
```diff
@@ -30,6 +30,14 @@ object Block {
       valueLength: Int = 90
   ): Block =
     Block.fromSource(Source.fromURI(path.toURI), keyLength, valueLength)
+
+  def sort(block: Block): Block = {
```
Comment (betarixm) [inline core/src/main/scala/Block.scala:34]: Should be defined in `class Block` with `def sort()`.

File: core/src/main/scala/Block.scala [hunk 5]
```diff
@@ -30,6 +30,14 @@ object Block {
...
+    val sortedRecords = block.records.sortBy(_.key)
+    new Block(sortedRecords)
+  }
+  def sample(block: Block): List[Key] = {
```
Comment (betarixm) [inline core/src/main/scala/Block.scala:38]: Should be defined in `class Block` with `def sample()`

File: core/src/main/scala/Block.scala (hunk 2 above)
Comment (betarixm) [reply core/src/main/scala/Block.scala:39]: e.g. 
```scala
  def sample(by: Int = 10): LazyList[Key] = {
    def sampleFromRecords(
        records: LazyList[Record],
        by: Int = by
    ): LazyList[Key] = {
      val (current, rest) = records.splitAt(by)
      val head = current.head.key
      head #:: sampleFromRecords(rest, by)
    }

    sampleFromRecords(records, by)
  }
```

File: core/src/main/scala/Block.scala (hunk 1 above)
Comment (hataehyeok) [reply core/src/main/scala/Block.scala:37]: Sure

File: core/src/main/scala/Block.scala (hunk 2 above)
Comment (hataehyeok) [reply core/src/main/scala/Block.scala]: Could I ask about the purpose of splitAt(interval)? It seems that splitAt(10) is intended to divide a list of Record objects into two parts at the specified index.
`Records -> (Records, Records)`
In the following line, we then obtain the key using the record.key.

From my perspective, splitAt(10)'s purpose is split Byte list to Key and dummy data like
`Byte -> (Key, dummy data)`
However, in this method, the argument is a list, not a byte

File: core/src/main/scala/Key.scala (hunk 3 above)
Comment (hataehyeok) [reply core/src/main/scala/Key.scala]: sure

File: core/src/test/scala/BlockSpec.scala [hunk 6]
```diff
@@ -97,4 +105,50 @@ class BlockSpec extends AnyFlatSpec {
...
+          new Record(new Key(Array(0x4)), Array(0x5, 0x6, 0x7)),
+          new Record(new Key(Array(0x8)), Array(0x9, 0xa, 0xb)),
+          new Record(new Key(Array(0xc)), Array(0xd, 0xe, 0xf))
+        )
+      )
+
+    assert(sortedBlock is expectedBlock)
+  }
+
+  it should "be able to sample" in {
```
[the remark this answers is not in the document]
Comment (hataehyeok) [reply core/src/test/scala/BlockSpec.scala]: ack

File: core/src/test/scala/KeySpec.scala [hunk 7]
```diff
@@ -20,4 +20,17 @@ class KeySpec extends AnyFlatSpec {
 
     assert(keyFromString is key)
   }
+
+  they should "be comparable" in {
+    val key1 = new Key(Array(0x01, 0x03, 0x03, 0x07) map (_.toByte))
+    val key2 = new Key(Array(0x01, 0x03, 0x03, 0x08) map (_.toByte))
+    val key3 = new Key(Array(0x01, 0x03, 0x03, 0x09) map (_.toByte))
+    val key4 = new Key(Array(0x01, 0x03, 0x03, 0x07) map (_.toByte))
+    val key5 = new Key(Array(0x01, 0x03, 0x01, 0x02) map (_.toByte))
```
[the remark this answers is not in the document]
Comment (hataehyeok) [reply core/src/test/scala/KeySpec.scala]: ack

File: core/src/main/scala/Block.scala (hunk 4 above)
Comment (hataehyeok) [reply core/src/main/scala/Block.scala]: ack

File: core/src/main/scala/Block.scala (hunk 5 above)
Comment (hataehyeok) [reply core/src/main/scala/Block.scala]: ack

File: core/src/main/scala/Block.scala (hunk 2 above)
Comment (betarixm) [reply core/src/main/scala/Block.scala]: I used `splitAt` to tie `n` keys and choose one. Other implementations welcomed. e.g. return `records.take(n)`
What's your idea?

File: core/src/main/scala/Block.scala [hunk 8]
```diff
@@ -53,4 +46,13 @@ class Block(val records: LazyList[Record]) extends AnyVal {
       // TODO: Handle exceptions
     } finally writer.close()
   }
+
+  def sort(block: Block): Block = {
```
Comment (betarixm) [inline core/src/main/scala/Block.scala:50]: Remove `block` parameter.

File: core/src/main/scala/Block.scala [hunk 9]
```diff
@@ -53,4 +46,13 @@ class Block(val records: LazyList[Record]) extends AnyVal {
...
+    val sortedRecords = block.records.sortBy(_.key)
+    new Block(sortedRecords)
+  }
+
+  def sample(block: Block): LazyList[Key] =
```
Comment (betarixm) [inline core/src/main/scala/Block.scala:55]: Remove `block` parameter.

File: core/src/main/scala/Block.scala (hunk 2 above)
Comment (betarixm) [reply core/src/main/scala/Block.scala]: FYI) `records.take(n)` could be harmful. Note that we couldn't know length of lazy list only if all elements are calcutated. That means, if we choose `records.take(n)`, we should take records in a rate (e.g. `10%`), then calculate all elements to get length, then take 10% of length.

Comment (hataehyeok) [reply core/src/main/scala/Block.scala]: I understand it! Thanks for the detailed explanation. There is an exception case that if the current is empty, we cannot bring it to the head of the current, so I added a base case for the recursion like
`    if (records.isEmpty)
      LazyList.empty[Key]`

File: core/src/main/scala/Block.scala (hunk 8 above)
Comment (hataehyeok) [reply core/src/main/scala/Block.scala]: ack

File: core/src/main/scala/Block.scala (hunk 9 above)
Comment (hataehyeok) [reply core/src/main/scala/Block.scala]: ack

**Review by betarixm — APPROVED** [review]

**Pull request merged**